### ISSUE TITLE
Can I use AddDbContextPool?

### ISSUE THREAD

**Issue by devalot76** [Finbuckle/Finbuckle.MultiTenant#375]
Hi and happy new year!
I was wondering, is it safe to use .AddDbContextPool() instead of .AddDbContext() or could there be issues with tenants since the first method reuses instances of the DbContext? 

My configuration is:
- .WithConfigurationStore()
- .WithRouteStrategy()
- .WithPerTenantAuthentication()

Thanks.

**Comment by AndrewTriesToCode** [comment]
hi, excellent question and I must admit I do not know off the top of my head. I will flag this and investigate. If you have experimented please let me know your findings.

**Comment by AndrewTriesToCode** [comment]
After looking into this I have to recommend against using the db context pool for exactly the reason you mentioned.

If you have a fancy db context that was smart enough to somehow validate its `ITenantInfo` member this could work but that is not something I'm planning to prioritize at the moment. Thanks again for the question.

**Comment by natelaff** [comment]
Would this apply to AddPooledDbContextFactory? Is this advisted against in all scenarios, or would this work in a db per tenant scenario?

**Comment by AndrewTriesToCode** [comment]
@natelaff I saw you had a comment on here.

It occurred to me the pooled context stuff might work if `IMultiTenantContextAccessor` is injected into the db context instead of `ITenantInfo` depending on the aspnet internals it might work -- would have to be tested.

**Comment by rohan-buechner** [comment]
Just a quick bump on this question. I see its been marked as stale, but as luck would have it, im in the same boat right now too. (Wanting to use HotChocolate (with a pooledbcontext) + Finbuckle)

I'll do some r&d into this as well, since its a bit of a core requirement for us, but I'm reaching out in the hopes that there has been some learnings since this question was created @AndrewTriesToCode 🙏  (Also major thank you for a great multi T lib, we've been using it for quite some time now and it works great.)

**Comment by AndrewTriesToCode** [comment]
hi, I unfortunately haven't tested out this use case yet so any help is appreciated. The main issue I see is how much of a db context  configuring is "locked in" during its initialization. If you are using the same connection string for all tenants and just changing the `TenantInfo` this very well could work since the global query filter is dynamic to check the `TenantId` for each query.

If tenants have different connection strings or even different database providers things get much more complicated. I assume some of the details are baked in during instantiation and can't be changed after the fact -- basically anything in your `OnConfiguring` method.

**Comment by natelaff** [comment]
Yeah this was where I tapped out of it and ran out of time. My connection strings are different per tenant. I would have loved to investigate more as this seems soooo much nicer than REST, but I am on an already delayed timeline :) Please share findings!

**Comment by g-6k** [comment]
Hello, I'm in the same situation. 
Want to use HotChocolate with Finbuckle and I really need PooledDbContext because of problem in DataLoader. 
Any new or plan to use Pooled Db Context?

**Comment by DylanSnel** [comment]
Hey guys,

I found a workaround

```cs
   [DataLoader]
   public static async Task<IReadOnlyDictionary<string, List<SiteVisitQuestionTemplate>>> GetSiteVisitQuestionTemplatesForTenantAsync(
       IReadOnlyList<string> tenantIds,
       IMultiTenantContextAccessor multiTenantContextAccessor,
       OnSiteContext context,
       CancellationToken ct)
   {
       context.TenantInfo = multiTenantContextAccessor.MultiTenantContext?.TenantInfo;

       return await context.SiteVisitQuestionTemplates
           .AsNoTracking()
           .Where(qt => tenantIds.Contains(qt.TenantId))
           .GroupBy(qt => qt.TenantId)
           .ToDictionaryAsync(g => g.Key, g => g.ToList(), ct);
   }
```

 Make sure you create a setter on your context

**Comment by DylanSnel** [comment]
I found an even better solution, disregard the previous

```cs
public class OnSiteContext(IMultiTenantContextAccessor multiTenantContextAccessor, DbContextOptions<OnSiteContext> options) : DbContext(options), IMultiTenantDbContext
{

   //... removed for brevity

    public ITenantInfo? TenantInfo { get => multiTenantContextAccessor.MultiTenantContext?.TenantInfo; }

   protected override void OnModelCreating(ModelBuilder modelBuilder)
   {
       base.OnModelCreating(modelBuilder);
       modelBuilder.ConfigureMultiTenant();

...

```

**Comment by devmaja** [comment]
Thanks, just ran into the same problem.
It might be worth mentioning in the docs.

If you use the extension method provided by Aspire (e.g.  AddNpgsqlDbContext from Aspire.Npgsql.EntityFrameworkCore.PostgreSQL) then the DbContext is always added with AddDbContextPool so this issue might come up more oftan in the future.

**Comment by AndrewTriesToCode** [comment]
> I found an even better solution, disregard the previous
> 
> public class OnSiteContext(IMultiTenantContextAccessor multiTenantContextAccessor, DbContextOptions<OnSiteContext> options) : DbContext(options), IMultiTenantDbContext
> {
> 
>    //... removed for brevity
> 
>     public ITenantInfo? TenantInfo { get => multiTenantContextAccessor.MultiTenantContext?.TenantInfo; }
> 
>    protected override void OnModelCreating(ModelBuilder modelBuilder)
>    {
>        base.OnModelCreating(modelBuilder);
>        modelBuilder.ConfigureMultiTenant();
> 
> ...

Yes this works well and I will likely add official support for this. However this introduces mutability to a context instance and can cause other problems in some situations so be careful.
